Enable Lowering Channels last Conv1x1 when max autotune is set

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #107004
* #106912
* #106911

This can lead to a large speedup when max autotune is set, e.g. resnet 2.1x -> 2.5x, particularly in combination with freezing. 


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

eellison committed:
[test]

[ghstack-poisoned]

eellison committed:
Update on "[test]"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

eellison committed:
Update on "Enable Lowering Channels last Conv1x1 when max autotune is set"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

eellison committed:
Update on "Enable Lowering Channels last Conv1x1 when max autotune is set"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

shunting314 commented:
I have a couple of questions.
1. could you explain more why the gain depends on enabling max-autotune for gemm? So unconditionally calling aten.mm for 1x1 conv does not result in the same gain?
2. is the speedup number '2.1x -> 2.5x' mentioned in the summary for training or inference? I assume it's inference but would like to double check.

eellison committed:
Update on "Enable Lowering Channels last Conv1x1 when max autotune is set"


This can lead to a large speedup when max autotune is set, e.g. resnet 2.1x -> 2.5x, particularly in combination with freezing. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

eellison commented:
> enabling max-autotune for gemm? So unconditionally calling aten.mm for 1x1 conv does not result in the same gain?

The mm shapes when you convert conv to 1x1 are pretty unusual. So I think the cublas heuristics are not as well tuned for them. For example, one addmm in resnet, `200704, 64` @ `64, 256`. Additionally, with max autotune we get to fuse the relu afterward which is a very cheap activation. 

> is the speedup number '2.1x -> 2.5x' mentioned in the summary for training or inference? I assume it's inference but would like to double check.

this is for inference. getting training data as well. but the pure mm shapes are faster, so it should help.

shunting314 commented on torch/_inductor/kernel/conv.py:
```diff
@@ -308,8 +308,20 @@ def convolution(
     dilation = pad_listlike(dilation, ndim)
     output_padding = pad_listlike(output_padding, ndim)
 
+    def channels_last_conv():
```
Any intuitive understanding for why this prefers channels last?

eellison replied:
Since we're permuting the input to `img.permute(0, 2, 3, 1)`, if the input is channels-last, it produces a dense tensor. If the input is contiguous, this produces a bad format for mm.

int3 commented on torch/_inductor/kernel/conv.py:
```diff
@@ -308,8 +308,20 @@ def convolution(
+        if V.graph.layout_opt and ndim == 2:
+            return True
+
+        layout = conv_layout(x, weight, None, **kwargs)
+        req_stride_order = ir.get_stride_order(
+            V.graph.sizevars.size_hints(layout.stride)
+        )
+        return req_stride_order == ir.NHWC_STRIDE_ORDER
```
just to make sure I understand -- this is checking that the layout of the output is NHWC, and if it is, then `x` and `w` are both NHWC?

eellison replied:
When we lower conv1x1 as mm, we are making the output NHWC. This is checking that we would have made it NHWC anyway, by looking at the output layout. 

When you call into `aten.convolution`, if either `x` or `w` is channels last, then the other will be made to be channels last as well with the op, then the kernel is invoked.

> x and w are both NHWC 

So, this is true with respect to the convolution kernel itself, but not necessarily the inputs to `aten::convolution`. Just one input needs to be NHWC for the output to be NHWC.

Within the conv1x1 as mm function, we force the input to be channels last strides. So we want to avoid :

- an extra copy to the input making it channels last where the copy wouldn't have happened otherwise
- accidentally changing the output striding, which might have downstream effects

eellison committed:
Update on "Enable Lowering Channels last Conv1x1 when max autotune is set"


This can lead to a large speedup when max autotune is set, e.g. resnet 2.1x -> 2.5x, particularly in combination with freezing. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

eellison commented:
@pytorchbot merge -f "unrelated failure"
